src/adm.rs

```diff
@@ -1,39 +1,225 @@
 use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap};
 use url::Url;
 
-use crate::error::HandlerError;
+use crate::error::{HandlerError, HandlerErrorKind, HandlerResult};
+use crate::server::ServerState;
+use crate::settings::Settings;
+//use crate::server::img_storage;
+
+pub(crate) const DEFAULT: &str = "DEFAULT";
 
 #[derive(Debug, Deserialize, Serialize)]
 pub struct AdmTileResponse {
     pub tiles: Vec<AdmTile>,
 }
 
-#[derive(Debug, Deserialize, Serialize)]
+/// Filter criteria for adm Tiles
+#[derive(Default, Clone, Debug)]
+pub struct AdmFilter {
+    /// list of allowed base host strings.
+    pub filter_set: BTreeMap<String, AdmAdvertiserFilterSettings>,
+}
+
+/// The AdmAdvertiserFilterSettings contain the settings for the various
+/// ADM provided partners. These are specified as a JSON formatted hash
+/// that contains the components. A special "DEFAULT" setting provides
+/// information that may be used as a DEFAULT, or commonly appearing set
+/// of data.
+/// See `impl From<Settings>` for details of the structure.
+#[derive(Clone, Debug, Deserialize, Default)]
+pub struct AdmAdvertiserFilterSettings {
+    pub(crate) advertiser_hosts: Vec<String>,
+    pub(crate) impression_hosts: Vec<String>,
+    pub(crate) position: Option<u8>,
+    pub(crate) include_regions: Vec<String>,
+}
+
+pub(crate) type AdmSettings = HashMap<String, AdmAdvertiserFilterSettings>;
+
+impl AdmFilter {
+    /// Check the impression URL to see if it's valid.
+    ///
+    /// This extends `filter_and_process`
+    fn check_impression(
```

Comment:
Likewise, the `click_url` can be checked as well.

Reply:
Cool, refactored things a bit to reduce possible code duplication.